### PR TITLE
Minibatch impl

### PR DESCRIPTION
**Pull request by Dahoas**
Implements minibatching for PPO. 

PPO sentiments `bs: 32, mbs: 16`. https://wandb.ai/dahoas/trlx/runs/oo6t8rla/overview?workspace=user-dahoas

PPO HH on GPT-NeoX `bs: 4, mbs: 1`. https://wandb.ai/dahoas/trlx/runs/9rnbmtu6?workspace=user-dahoas

### CODE REVIEW

File: trlx/data/configs.py
Comment (cat-state) [inline trlx/data/configs.py:200]: (very pedantic) nit, feel free to ignore: usually I've heard this called _micro_ batch, with minibatch referring to what we usually call "a batch" (to distinguish from a single batch of the whole dataset)

File: trlx/trainer/accelerate_base_trainer.py
Comment (cat-state) [inline trlx/trainer/accelerate_base_trainer.py:488]: Shouldn't this get loss using a minibatch sliced from `batch`?

Comment (Dahoas) [reply trlx/trainer/accelerate_base_trainer.py:488]: Ah yeah good catch (I copied this over from a different branch and forgot to change this)

File: trlx/trainer/accelerate_base_trainer.py
Comment (eluzhnica) [inline trlx/trainer/accelerate_base_trainer.py:497]: to avoid unnecessary gradient synchronization when doing using gradient accumulation you can simply add: `self.accelerator.accumulate(self.model)` here

Comment (eluzhnica) [inline trlx/trainer/accelerate_base_trainer.py:497]: to avoid unnecessary gradient synchronization when doing using gradient accumulation you can simply add: `self.accelerator.accumulate(self.model)` here

Comment (cat-state) [reply trlx/trainer/accelerate_base_trainer.py:497]: fwiw i tried using that atop this PR and got weird results: https://wandb.ai/uwu1/trlx/reports/Untitled-Report--VmlldzozOTAyNDg4

**Review by cat-state — APPROVED** [review]
LGTM!

File: trlx/trainer/accelerate_base_trainer.py
Comment (eluzhnica) [reply trlx/trainer/accelerate_base_trainer.py:497]: yes, but we're specifying that already in the config files like zero2-bf16.yaml

did you add it right below this line or literally at the top of the script? it should be added below this line since the entering of the context is how `accumulate` tracks the number of steps that are being executed so that it knows when to sync (has an internal step counter)

Also the division of the loss by self.num_mb should go away since that would be handled by accelerator

Here's an example: https://github.com/muellerzr/timing_experiments/blob/main/good.py#L152

Comment (eluzhnica) [reply trlx/trainer/accelerate_base_trainer.py:497]: @cat-state Okay just made a [PR](https://github.com/CarperAI/trlx/pull/396). I'll try to run the same tests that @Dahoas ran to confirm it works as intended.

As for the TRLConfig, happy to do so, but it seems to me that the configs for accelerate are being set up separately in the repo (configs/accelerate/....) and that is the general pattern I've seen before too. And accelerate feeds those params behind the scenes for us automatically, so if we were to also specify in the TRLConfig it is a bit redundant (and potentially conflictual values). let me know what you think

File: trlx/trainer/accelerate_base_trainer.py
Comment (eluzhnica) [inline trlx/trainer/accelerate_base_trainer.py:479]: @Dahoas Just to confirm, this is still a draft correct? Since (although I haven't run it) this I think would break for ILQLBatch and other types of datatypes

Comment (cat-state) [reply trlx/trainer/accelerate_base_trainer.py:479]: Ah you're right, I just ran this with the benchmarking scripts and it crashed here with this error

**Review by cat-state — CHANGES_REQUESTED** [review]
Need to make it not incompatible with ILQL

File: trlx/trainer/accelerate_base_trainer.py
Comment (eluzhnica) [reply trlx/trainer/accelerate_base_trainer.py:479]: Fixed this here: https://github.com/CarperAI/trlx/pull/403

**Comment by Dahoas** [comment]
Merge #403 which makes mini-batching compatible with all trainers.

**Comment by Dahoas** [comment]
Let's merge this into main once #396 gets merged.

**Comment by Dahoas** [comment]
@cat-state Can you take another look and if it looks good we can merge?

**Comment by cat-state** [comment]
thanks @eluzhnica and @Dahoas ! just tried w ilql and it seems to work https://wandb.ai/uwu1/trlx/runs/m2e4rwga?workspace=user-uwu1
Thanks for working together and adding this!

**Review by cat-state — APPROVED** [review]

**Pull request merged**
